addons/editor-theme3/theme3.js

```diff
@@ -1,398 +1,282 @@
+  Blockly.FieldVariableGetter.prototype.init = function () {
+    // Variable reporters
+    oldFieldVariableGetterInit.call(this);
+    this.textElement_.style.fill = textColor(this);
+  };
+
+  const oldFieldMatrixUpdateMatrix = Blockly.FieldMatrix.prototype.updateMatrix_;
+  Blockly.FieldMatrix.prototype.updateMatrix_ = function () {
+    // Matrix inputs
+    oldFieldMatrixUpdateMatrix.call(this);
+    console.log("updating matrix");
+    console.log("still updating matrix");
+    for (let i = 0; i < this.matrix_.length; i++) {
+      if (this.matrix_[i] !== "0") {
+        this.fillMatrixNode_(this.ledButtons_, i, uncoloredTextColor());
+        this.fillMatrixNode_(this.ledThumbNodes_, i, uncoloredTextColor());
       }
     }
-  }
+  };
 
-  document.documentElement.style.setProperty(
-    "--editorTheme3-inputColor-text",
-    textColor(addon.settings.get("input-color"))
-  );
-  newStyle.textContent = stylesheet;
-}
+  const oldFieldMatrixCreateButton = Blockly.FieldMatrix.prototype.createButton_;
+  Blockly.FieldMatrix.prototype.createButton_ = function (fill) {
+    if (fill === "#FFFFFF") fill = uncoloredTextColor();
+    return oldFieldMatrixCreateButton.call(this, fill);
+  };
+
+  const updateColors = () => {
+    const workspace = Blockly.getMainWorkspace();
+    const toolbox = workspace.getToolbox();
+    const vm = addon.tab.traps.vm;
 
-export default async function ({ addon, global, console }) {
-  const otherStyle = document.querySelector(`[data-addon-id='${addon.self.id}']`);
-  const newStyle = document.createElement("style");
-  updateSettings(addon, newStyle);
-  addon.settings.addEventListener("change", () => {
-    updateSettings(addon, newStyle);
-  });
-  newStyle.className = "scratch-addons-style";
-  newStyle.setAttribute("data-addon-id", addon.self.id);
-  newStyle.setAttribute("data-addon-index", otherStyle.getAttribute("data-addon-index"));
+    textMode = addon.settings.get("text");
 
-  document.documentElement.insertBefore(newStyle, otherStyle.nextSibling);
+    for (let category of categories) {
+      Blockly.Colours[category.colorId].primary = primaryColor(category);
+      Blockly.Colours[category.colorId].secondary = secondaryColor(category);
+      Blockly.Colours[category.colorId].tertiary = tertiaryColor(category);
+    }
+    Blockly.Colours.textField = otherColor("input-color", "textField");
+    if (uncoloredTextColor() === "#575e75") Blockly.Colours.fieldShadow = "rgba(0, 0, 0, 0.15)";
+
+    // Reload toolbox
+    toolbox.dispose();
+    toolbox.workspace_ = workspace;
+    toolbox.init();
+    // Connects events to VM
+    // https://github.com/LLK/scratch-gui/blob/ba76db7350bd43b79119cac2701bc10f6c511f0c/src/containers/blocks.jsx#L250-L254
+    const flyoutWorkspace = workspace.getFlyout().getWorkspace();
+    flyoutWorkspace.addChangeListener(vm.flyoutBlockListener);
+    flyoutWorkspace.addChangeListener(vm.monitorBlockListener);
```

Comment:
like in this commit, there should specific things you want to reload, instead of disposing and initing, which probably is slower. plus, its probably better not to touch vm here.

Reply:
Is it possible to redraw every block in the flyout without recreating the workspace (which requires interacting with the VM)?

Reply:
@mxmou How does this sound... I tested it briefly and the flyout still seemed to work.
```suggestion
    toolbox.flyout_.dispose();
    toolbox.createFlyout_();
    toolbox.showAll_();
    toolbox.flyout_.position();
```

Reply:
There is still a bug: category bubbles are not updating.